### PR TITLE
TES-305 DNS Module

### PR DESCRIPTION
**Pull request by kristianiliev1**
## Description

Introduced new Private DNS Module

## Related Issues

[TES-305](https://ontotext.atlassian.net/browse/TES-305)

## Changes
Added Private DNS Module
Added script, which creates DNS records based on the attached disk
Changed disk naming to "Disk-$${RESOURCE_GROUP}-$${ZONE_ID}-$${DISK_ORDER}"
Fixed an issue where the attach disk script was failing

## Checklist

- [x] I have tested these changes thoroughly.
- [x] My code follows the project's coding style.
- [x] I have added appropriate comments to my code, especially in complex areas.
- [x] All new and existing tests passed locally.

[TES-305]: https://ontotext.atlassian.net/browse/TES-305?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -152,3 +152,22 @@ module "vm" {
...
   ]
 }
+
+
+module "dns" {
+  source               = "./modules/dns"
+  resource_name_prefix = var.resource_name_prefix
+  resource_group_name  = var.resource_group_name
+  zone_dns_name        = var.zone_dns_name
+  # principal_id         = module.identity.identity_name
```
Comment (yaskoo) [inline main.tf:162]: don't leave commented out stuff

File: main.tf [hunk 2]
```diff
@@ -152,3 +152,22 @@ module "vm" {
...
+  # virtual_network_id   = azurerm_virtual_network.graphdb.id
+  identity_name = module.identity.identity_name
+  depends_on = [
```
Comment (yaskoo) [inline main.tf:165]: We should discuss the use of `depends_on` + data sources. HashiCorn [discourage](https://developer.hashicorp.com/terraform/language/meta-arguments/depends_on) the use of `depends_on`, unless absolutely necessary and it seems that the use of data sources should be for things that are [assumed to exist](https://support.hashicorp.com/hc/en-us/articles/4414363057299-Best-Practice-depends-on-Meta-Argument).

File: main.tf [hunk 3]
```diff
@@ -152,3 +152,22 @@ module "vm" {
...
+    azurerm_resource_group.graphdb,
+    azurerm_virtual_network.graphdb,
+    azurerm_subnet.graphdb-vmss,
+    module.configuration,
+    module.identity
+
+  ]
+}
```
Comment (yaskoo) [inline main.tf:173]: add empty line

Comment (kristianiliev1) [reply main.tf:173]: Added

File: main.tf (hunk 1 above)
Comment (kristianiliev1) [reply main.tf:162]: Understood.

File: main.tf [hunk 4]
```diff
@@ -152,3 +152,22 @@ module "vm" {
     module.configuration
   ]
 }
+
+
+module "dns" {
+  source               = "./modules/dns"
+  resource_name_prefix = var.resource_name_prefix
+  resource_group_name  = var.resource_group_name
```
[the remark this answers is not in the document]
Comment (kristianiliev1) [reply main.tf:160]: I will test and update later.

File: modules/dns/main.tf [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+resource "azurerm_private_dns_zone_virtual_network_link" "zone_link" {
+  name                  = "${var.resource_name_prefix}-dns-link"
+  resource_group_name   = data.azurerm_resource_group.graphdb.name
+  private_dns_zone_name = azurerm_private_dns_zone.zone.name
+  virtual_network_id    = data.azurerm_virtual_network.graphdb.id
+}
+
+resource "azurerm_role_assignment" "dns_zone_role_assignment" {
+  principal_id         = data.azurerm_user_assigned_identity.graphdb-instances.principal_id #from identity module
```
[the remark this answers is not in the document]
Comment (kristianiliev1) [reply modules/dns/main.tf:28]: Removed.

File: modules/dns/main.tf [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  name                = var.identity_name
+  resource_group_name = var.resource_group_name
+}
+
+resource "azurerm_private_dns_zone" "zone" {
+  name                = "${var.resource_name_prefix}.dns.zone"
+  resource_group_name = var.resource_group_name
+}
+
+data "azurerm_virtual_network" "graphdb" {
```
[the remark this answers is not in the document]
Comment (kristianiliev1) [reply modules/dns/main.tf:15]: Moved.

File: modules/dns/variables.tf [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  description = "DNS name for the private DNS zone in Azure"
+  type        = string
+}
+
+variable "resource_group_name" {
+  description = "Resource group name where the DNS zone will be created"
+  type        = string
+}
+
+# variable "virtual_network_id" {
```
[the remark this answers is not in the document]
Comment (kristianiliev1) [reply modules/dns/variables.tf:11]: Removed

File: modules/dns/variables.tf [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+#   description = "ID of the virtual network where the DNS zone will be linked"
```
[the remark this answers is not in the document]
Comment (kristianiliev1) [reply modules/dns/variables.tf:12]: Deleted

File: variables.tf [hunk 9]
```diff
@@ -93,3 +93,24 @@ variable "custom_graphdb_vm_user_data" {
   type        = string
   default     = null
 }
+
+variable "resource_group_name" {
+  description = "Resource group name"
+  type        = string
+}
+
+# variable "principal_id" {
```
[the remark this answers is not in the document]
Comment (kristianiliev1) [reply variables.tf:102]: Removed

File: variables.tf [hunk 10]
```diff
@@ -93,3 +93,24 @@ variable "custom_graphdb_vm_user_data" {
...
+
+# variable "virtual_network_id" {
+#   description = "Virtual network ID"
+#   type        = string
+# }
+
+variable "zone_dns_name" {
+  description = "Zone DNS Name"
+  type        = string
+}
```
[the remark this answers is not in the document]
Comment (kristianiliev1) [reply variables.tf:116]: Added

File: modules/dns/variables.tf [hunk 11]
```diff
@@ -0,0 +1,29 @@
+variable "zone_dns_name" {
```
[the remark this answers is not in the document]
Comment (kristianiliev1) [reply modules/dns/variables.tf:1]: removed after consultation

File: modules/load_balancer/main.tf [hunk 12]
```diff
@@ -27,7 +27,7 @@ resource "azurerm_public_ip" "graphdb-load-balancer" {
   allocation_method = "Static"
   zones             = var.zones
   # TODO: Should be provided and if not, this to be the default
-  domain_name_label = "${var.resource_name_prefix}-${random_string.fqdn.result}"
+  domain_name_label = "${lower(var.resource_name_prefix)}-${random_string.fqdn.result}"
```
[the remark this answers is not in the document]
Comment (kristianiliev1) [reply modules/load_balancer/main.tf:30]: Agreed. Adding validation.

**Comment by viktor-ribchev** [comment]
Rebase and remove data sources in favor of variables.

File: .gitignore [hunk 13]
```diff
@@ -35,3 +35,9 @@ terraform.rc
 
 # IDEs
 .idea/
+
+#Certificates
+cerificate.*
+privatekey.pem
+certificate.pem
```
Comment (mihailradkov) [inline .gitignore:42]: Simplify the conditions, the last two are already covered. On the other hand, using the workdir for such files should be avoided

File: modules/dns/versions.tf [hunk 14]
```diff
@@ -0,0 +1,10 @@
+terraform {
```
Comment (mihailradkov) [inline modules/dns/versions.tf:1]: No longer needed for sub modules. Remove this file.

File: modules/dns/variables.tf [hunk 15]
```diff
@@ -0,0 +1,35 @@
+variable "resource_group_name" {
+  description = "Resource group name where the DNS zone will be created"
+  type        = string
+}
+
+variable "resource_name_prefix" {
+  description = "Resource name prefix used for tagging and naming Azure resources"
+  type        = string
+
+  validation {
```
[the remark this answers is not in the document]
Comment (viktor-ribchev) [reply modules/dns/variables.tf]: done

**Review by yaskoo — APPROVED** [review]

**Pull request merged**